dmoj/cptbox/ptdebug_arm64.cpp

```diff
@@ -3,109 +3,109 @@
 }
 
-long pt_debugger_arm64::peek_reg(int reg) {
-    return arm64_reg[reg];
-}
-
-void pt_debugger_arm64::poke_reg(int reg, long data) {
-    arm64_reg[reg] = data;
-    arm_reg_changed = true;
-}
-
-int pt_debugger_arm64::syscall() {
-    return (int) peek_reg(ARM_x8);
-}
-
-long pt_debugger_arm64::result() {
-    return peek_reg(ARM_x0);
-}
-
-void pt_debugger_arm64::result(long value) {
-    poke_reg(ARM_x0, value);
+#define INVALID_ABI(source) fprintf(stderr, source ": Invalid ABI\n")
+
+int pt_debugger::syscall() {
+    switch (abi_) {
+        case PTBOX_ABI_ARM:
+            return regs.arm32.r7;
+        case PTBOX_ABI_ARM64:
+            return regs.arm64.regs[8];
+        default:
+            INVALID_ABI("ptdebug_arm64.cpp:syscall getter");
+            return -1;
+    }
 }
 
-long pt_debugger_arm64::arg0() {
-    return peek_reg(ARM_x0);
-}
+void pt_debugger::syscall(int id) {
+    struct iovec iovec;
+    iovec.iov_base = &id;
+    iovec.iov_len = sizeof id;
 
-void pt_debugger_arm64::arg0(long data) {
-#if !PTBOX_FREEBSD
-    if (is_enter())
-        poke_reg(ARM_x0, data);
-#endif
+    if (ptrace(PTRACE_SETREGSET, tid, NT_ARM_SYSTEM_CALL, &iovec))
+        perror("ptrace(PTRACE_SETREGSET, NT_ARM_SYSTEM_CALL)");
 }
 
-#define make_arg(id, reg) \
-    long pt_debugger_arm64::arg##id() { \
-        return peek_reg(reg); \
+#define MAKE_ACCESSOR(method, arm32_name, arm64_name) \
+    long pt_debugger::method() { \
+        switch (abi_) { \
+            case PTBOX_ABI_ARM: \
+                return regs.arm32.arm32_name; \
+            case PTBOX_ABI_ARM64: \
+                return regs.arm64.arm64_name; \
+            default: \
+                return INVALID_ABI("ptdebug_arm64.cpp:" #method " getter"); \
+        } \
     } \
     \
-    void pt_debugger_arm64::arg##id(long data) {\
-        poke_reg(reg, data); \
+    void pt_debugger::method(long value) { \
+        regs_changed = true; \
+        switch (abi_) { \
+            case PTBOX_ABI_ARM: \
+                regs.arm32.arm32_name = value; \
+                return; \
+            case PTBOX_ABI_ARM64: \
+                regs.arm64.arm64_name = value; \
+                return; \
+            default: \
+                INVALID_ABI("ptdebug_arm64.cpp:" #method " setter"); \
+        } \
     }
 
-make_arg(1, ARM_x1);
-make_arg(2, ARM_x2);
-make_arg(3, ARM_x3);
-make_arg(4, ARM_x4);
-make_arg(5, ARM_x5);
-
-#undef make_arg
+MAKE_ACCESSOR(result, r7, regs[8])
+MAKE_ACCESSOR(arg0, r0, regs[0])
+MAKE_ACCESSOR(arg1, r1, regs[1])
+MAKE_ACCESSOR(arg2, r2, regs[2])
+MAKE_ACCESSOR(arg3, r3, regs[3])
+MAKE_ACCESSOR(arg4, r4, regs[4])
+MAKE_ACCESSOR(arg5, r5, regs[5])
 
-bool pt_debugger_arm64::is_exit(int syscall) {
-    return syscall == 93 || syscall == 94;
-}
-
-int pt_debugger_arm64::getpid_syscall() {
-    return 172;
-}
+#undef MAKE_ACCESSOR
 
-pt_debugger_arm64::pt_debugger_arm64() {
-#if PTBOX_SECCOMP
-    execve_id = 221;
-#else
-    // execve is actually 221, but...
```

Comment:
Maybe keep this comment line

Reply:
Do we need to keep that line? It's obvious in the `return` that `execve` is 221.